### ISSUE TITLE
Types with module-level visibility in public Types

### ISSUE THREAD

**Issue by sameerajayasoma** [ballerina-platform/ballerina-spec#168]
Consider the following code snippet. Since Ballerina uses structural equivalence to define type equivalence, the following scenario is valid in most cases.   

```ballerina
// This is a user-defined type with module-level visibility
type T ...;
public function doSomething(T t) {}
```
But this may not be valid if T is a subtype of `object {}`.  And for these object types, we should even stop people make them public. e.g. `public type T ...;`

Should we define this as part of the language spec?

**Comment by jclark** [comment]
By "not valid", I am guessing you mean that it may not be possible for another module to call this function because

1. T has private or protected members that make it impossible to define a subtype of T in another module, and
2. The module defining T does not define any public subtypes of T

Is that right?

**Comment by sameerajayasoma** [comment]
That is correct, James. I think we can make the 2nd point more precious as below.

2. T has no private members and also the module defining T does not define any public subtypes of T.

**Comment by sameerajayasoma** [comment]
I used a public function as an example. But it can be any public Type

```ballerina
<<module_x>>
type T ...;
public type U record {
     T t;
}
```

In some situations as described in previous comments, we have to use name equivalence for type T. Since T is not visible to other modules, those modules cannot use the record type U.
